terraform-provider-constellation/internal/provider/image_data_source.go

```diff
@@ -0,0 +1,167 @@
+					"See the [full list of CSPs](https://docs.edgeless.systems/constellation/overview/clouds) that Constellation supports.",
+				Required: true,
+				Validators: []validator.String{
+					stringvalidator.OneOf("aws", "azure", "gcp"),
+				},
+			},
+			"region": schema.StringAttribute{
+				Description: "Region to retrieve the image for. Only required for AWS.",
+				MarkdownDescription: "Region to retrieve the image for. Only required for AWS.\n" +
+					"The Constellation OS image must be [replicated to the region](https://docs.edgeless.systems/constellation/workflows/config)," +
+					"and the region must [support AMD SEV-SNP](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/snp-requirements.html), if it is used for Attestation.",
+				Optional: true,
+			},
+			"reference": schema.StringAttribute{
+				Description:         "CSP-specific reference to the image.",
+				MarkdownDescription: "CSP-specific reference to the image.",
+				Computed:            true,
+			},
+		},
+	}
+}
+
+// TODO(msanft): Possibly implement more complex validation for inter-dependencies between attributes.
+// E.g., region should be required if, and only if, AWS is used.
+
+// Configure configures the data source.
+func (d *ImageDataSource) Configure(_ context.Context, _ datasource.ConfigureRequest, _ *datasource.ConfigureResponse) {
+	// Create the image-fetcher client.
+	d.imageFetcher = imagefetcher.New()
+}
+
+// Read reads from the data source.
+func (d *ImageDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	// Retrieve the configuration values for this data source instance.
+	var data ImageDataSourceModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	// Check configuration for errors.
+	csp := cloudprovider.FromString(data.CSP.ValueString())
+	if csp == cloudprovider.Unknown {
+		resp.Diagnostics.AddAttributeError(
+			path.Root("csp"),
+			"Invalid CSP",
+			fmt.Sprintf("Invalid CSP: %s", data.CSP.ValueString()),
+		)
+	}
+
+	attestationVariant, err := variant.FromString(data.AttestationVariant.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddAttributeError(
+			path.Root("attestation_variant"),
+			"Invalid Attestation Variant",
+			fmt.Sprintf("When parsing the Attestation Variant (%s), an error occurred: %s", data.AttestationVariant.ValueString(), err),
+		)
+	}
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Retrieve Image Reference
+	imageRef, err := d.imageFetcher.FetchReference(ctx, csp, attestationVariant, data.ImageVersion.ValueString(), data.Region.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error fetching Image Reference",
+			fmt.Sprintf("When fetching the image reference, an error occurred: %s", err),
+		)
+		return
+	}
+
+	// Save data into Terraform state
+	data.Reference = types.StringValue(imageRef)
+	// Use a placeholder ID for testing, as per
+	data.ID = types.StringValue("placeholder")
```

Comment:
is this relevant?

Reply:
Yes. see https://developer.hashicorp.com/terraform/plugin/framework/acctests#no-id-found-in-attributes